lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -665,14 +678,140 @@ def write(chunk)
+                partial_error_count = indexes.length
+                increment_dropped_entries_count(partial_error_count)
+                @log.warn "Dropping #{partial_error_count} log message(s)",
+                          error_code: "google.rpc.Code.#{error_code}",
+                          error: error_message
+              end
+            end
           end
         end
       end
     end
 
+    # Extract a map of error details from an potentially partially successful
+    # request. The keys in this map are the indexes of the failed log entries in
+    # the original json payload. The values in this map are the error code and
+    # error message.
+    #
+    # A sample error.body looks like:
+    # {
+    #   "error": {
+    #     "code": 403,
+    #     "message": "User not authorized.",
+    #     "status": "PERMISSION_DENIED",
+    #     "details": [
+    #       {
+    #         "@type": "type.googleapis.com/google.logging.v2.WriteLogEntriesPar
+    #           tialErrors",
+    #         "logEntryErrors": {
+    #           "0": {
+    #             "code": 7,
+    #             "message": "User not authorized."
+    #           },
+    #           "1": {
+    #             "code": 3,
+    #             "message": "Log name contains illegal character :"
+    #           }
+    #         }
+    #       },
+    #       {
+    #         "@type": "type.googleapis.com/google.rpc.DebugInfo",
+    #         "detail": "[ORIGINAL ERROR] generic::permission_denied: User not a
+    #           uthorized. [google.rpc.error_details_ext] { message: \"User not
+    #           authorized.\" details { type_url: \"type.googleapis.com/google.l
+    #           ogging.v2.WriteLogEntriesPartialErrors\" value: \"\\n\\034\\010\
+    #           \000\\022\\030\\010\\007\\022\\024User not authorized.\\n-\\010\
+    #           \001\\022)\\010\\003\\022%Log name contains illegal character :\
+    #           " } }"
+    #       }
+    #     ]
+    #   }
+    # }
+    #
+    # In this exxample, the logEntryErrors we are extracting should be:
+    # {
+    #   "0": {
+    #     "code": 7,
+    #     "message": "User not authorized."
+    #   },
+    #   "1": {
+    #     "code": 3,
+    #     "message": "Log name contains illegal character :"
+    #   }
+    # }
+    def extract_log_entry_errors(error)
+      error_details = parse_json_or_nil(error.body)['error']['details']
+      partial_errors = error_details.detect do |error_detail|
+        error_detail['@type'] == PARTIAL_ERROR_KEY
+      end
+      partial_errors['logEntryErrors']
+    rescue => e
+      @log.warn 'Failed to extract log entry errors from the error details' \
+                " json: #{error.body}.", error: e
+      {}
+    end
+
+    # Given the logEntryErrors, constuct a map from errors to a list of indexes
+    # of log entries that failed for this specific error.
+    #
+    # The logEntryErrors should be in the format of:
```

Comment:
I wouldn't bother repeating this here -- maybe construct one example that shows the full details, the extracted errors, and the resulting map, and put it above these two methods?

Reply:
Removed the duplicated example and changed the phrasing to:

```
    # For logEntryErrors as in the example above (in the comment section of the
    # function extract_log_entry_errors), the constructed map would be:
    # {
    #   [
    #     7,
    #     'User not authorized.'
    #   ]: ["0"],
    #   [
    #     3,
    #     'Log name contains illegal character :'
    #   ]: ["1", "2"]
    # }
```

Reply:
Careful, the example above only had 2 error indexes, and in your map you have 3. Should we update the example to have two illegal character errors?

Reply:
Meant to do that. Thanks for catching this.